storage/src/main/java/tech/pegasys/teku/storage/client/BlockTimelinessTracker.java

```diff
@@ -87,7 +94,34 @@ public void setBlockTimelinessFromArrivalTime(
             });
   }
 
-  public Optional<Boolean> isBlockTimely(final Bytes32 root) {
+  Optional<Boolean> isBlockTimely(final Bytes32 root) {
     return Optional.ofNullable(blockTimeliness.get(root));
   }
+
+  // is_proposing_on_time from consensus-spec
+  // 'on time' is before we're half-way to the attester time. logically, if the slot is 3 segments,
+  // then splitting into 6 segments is half-way to the attestation time.
+  public boolean isProposingOnTime(final UInt64 slot) {
+    final UInt64 slotStartTimeMillis =
+        spec.getSlotStartTimeMillis(slot, genesisTimeMillisSupplier.get());
+    final UInt64 timelinessLimit = spec.getMillisPerSlot(slot).dividedBy(INTERVALS_PER_SLOT * 2);
+    final UInt64 currentTimeMillis = timeProvider.getTimeInMillis();
+    final boolean isTimely =
+        currentTimeMillis.minusMinZero(slotStartTimeMillis).isLessThan(timelinessLimit);
```

Comment:
In the spec, I noticed the current time is coming from the store (`store.time`). Is the current time from the time provider equivalent or should we be using `Store.getTimeMillis()`?

The same question applies to store.genesis_time and `Store.getGenesisTime()`.

Reply:
genesisTime is memoized, we don't need to check continually...

store time is reliant on other factors, i think we want real time if we're talking about potentially doing a reorg... it'll be a much more accurate indication of how much time we have